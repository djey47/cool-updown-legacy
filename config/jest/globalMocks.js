// JS API //
const mockDateNow = jest.fn(() => new Date('June 12, 2018 14:14:00'));
Date.now = mockDateNow;

// NODE MODULES //
// app-root-dir
const mockAppRootDirGet = jest.fn(() => './test');
jest.mock('app-root-dir', () => ({
  get: () => mockAppRootDirGet(),
}));

// cron
const mockCronStart = jest.fn();
jest.mock('cron', () => ({
  CronJob: jest.fn(() => ({
    start: () => mockCronStart(),
  })),
}));

// express-basic-auth
const mockBasicAuthPlugin = jest.fn();
jest.mock('express-basic-auth', () =>
  settings => mockBasicAuthPlugin(settings));

// node-ssh
const mockSSHConnect = jest.fn();
const mockSSHExecCommand = jest.fn();
const mockSSHDispose = jest.fn();
jest.mock('node-ssh', () => jest.fn(() => ({
  connect: o => mockSSHConnect(o),
  execCommand: c => mockSSHExecCommand(c),
  dispose: () => mockSSHDispose(),
})));

// shelljs
const mockShellMkdir = jest.fn();
jest.mock('shelljs', () => ({
  mkdir: (o, p) => mockShellMkdir(o, p),
}));

// wake_on_lan
const mockWOLWake = jest.fn();
jest.mock('wake_on_lan', () => ({
  wake: (a, o, f) => mockWOLWake(a, o, f),
}));

// winston
const mockWinstonLogger = jest.fn(() => ({
  error: () => null,
  log: () => null,
}));
const mockWinstonTransport = jest.fn();
jest.mock('winston', () => ({
  Logger: mockWinstonLogger,
  transports: {
    Console: mockWinstonTransport,
    File: mockWinstonTransport,
  },
}));

module.exports = {
  dateMock: {
    now: mockDateNow,
  },
  appRootDirMock: {
    get: mockAppRootDirGet,
  },
  cronMock: {
    start: mockCronStart,
  },
  expressBasicAuthMock: mockBasicAuthPlugin,
  nodesshMock: {
    connect: mockSSHConnect,
    execCommand: mockSSHExecCommand,
    dispose: mockSSHDispose,
  },
  shelljsMock: {
    mkdir: mockShellMkdir,
  },
  winstonMock: {
    Logger: mockWinstonLogger,
    Transport: mockWinstonTransport,
  },
  wakeonlanMock: {
    wake: mockWOLWake,
  },
};
