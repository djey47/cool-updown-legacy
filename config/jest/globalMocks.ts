// JS API //
const mockDateNow = jest.fn(() => new Date('June 12, 2018 14:14:00Z'));
Date.now = mockDateNow as unknown as () => number;

// NODE API //
// fs
const mockFSReadFile = jest.fn();
jest.mock('fs/promises', () => ({
  readFile: async (p, e) => mockFSReadFile(p, e),
}));

// NODE MODULES //
// app-root-dir
const mockAppRootDirGet = jest.fn(() => './test');
jest.mock('app-root-dir', () => ({
  get: () => mockAppRootDirGet(),
}));

// axios
const mockAxiosGet = jest.fn();
jest.mock('axios', () => ({
  get: url => mockAxiosGet(url),
}));

// cron
const mockCronStart = jest.fn();
const mockCronJob = jest.fn(() => ({
  start: () => mockCronStart(),
}));
jest.mock('cron', () => ({
  CronJob: mockCronJob,
}));

// express-basic-auth
const mockBasicAuthPlugin = jest.fn();
jest.mock('express-basic-auth', () => settings => mockBasicAuthPlugin(settings));

// node-ssh
const mockSSHConnect = jest.fn();
const mockSSHExecCommand = jest.fn();
const mockSSHDispose = jest.fn();
jest.mock('node-ssh', () => ({
  NodeSSH: jest.fn(() => ({
    connect: o => mockSSHConnect(o),
    execCommand: c => mockSSHExecCommand(c),
    dispose: () => mockSSHDispose(),
  })),
}));

// shelljs
const mockShellMkdir = jest.fn();
jest.mock('shelljs', () => ({
  mkdir: (o, p) => mockShellMkdir(o, p),
}));

// wake_on_lan
const mockWOLWake = jest.fn();
jest.mock('wake_on_lan', () => ({
  wake: (a, o, f) => mockWOLWake(a, o, f),
}));

// winston
const mockWinstonLogger = jest.fn(() => ({
  error: () => null,
  log: () => null,
}));
const mockWinstonTransport = jest.fn();
jest.mock('winston', () => ({
  Logger: mockWinstonLogger,
  transports: {
    Console: mockWinstonTransport,
    File: mockWinstonTransport,
  },
}));

// PROJECT SPECIFIC //
const mockExpressResponseStatus = jest.fn();
const mockExpressResponseSend = jest.fn();
const mockExpressResponseJson = jest.fn();

const mockOnJobStart = jest.fn();
const mockOffJobStart = jest.fn();
const mockOnJobStop = jest.fn();
const mockOffJobStop = jest.fn();

const mockGatewayPing = jest.fn();

export default {
  dateMock: {
    now: mockDateNow,
  },
  appRootDirMock: {
    get: mockAppRootDirGet,
  },
  axiosMock: {
    get: mockAxiosGet,
  },
  cronMock: {
    Job: mockCronJob,
    start: mockCronStart,
  },
  expressBasicAuthMock: mockBasicAuthPlugin,
  nodesshMock: {
    connect: mockSSHConnect,
    execCommand: mockSSHExecCommand,
    dispose: mockSSHDispose,
  },
  shelljsMock: {
    mkdir: mockShellMkdir,
  },
  winstonMock: {
    Logger: mockWinstonLogger,
    Transport: mockWinstonTransport,
  },
  wakeonlanMock: {
    wake: mockWOLWake,
  },
  expressResponseMock: {
    statusMock: mockExpressResponseStatus,
    sendMock: mockExpressResponseSend,
    jsonMock: mockExpressResponseJson,
    status: c => mockExpressResponseStatus(c),
    send: msg => mockExpressResponseSend(msg),
    json: () => mockExpressResponseJson(),
  },
  jobsMock: {
    onJobStart: mockOnJobStart,
    offJobStart: mockOffJobStart,
    onJobStop: mockOnJobStop,
    offJobStop: mockOffJobStop,
  },
  systemGatewayMock: {
    ping: h => mockGatewayPing(h),
    pingMock: mockGatewayPing,
  },
  nodeFSMock: {
    readFile: mockFSReadFile,
  },
};
