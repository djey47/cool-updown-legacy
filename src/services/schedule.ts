import { validateInputParameters } from '../helpers/api';
import logger from '../helpers/logger';
import { TypedResponse } from '../model/express';
import { ApiRequest, AppState, ServerState } from '../model/models';
import messages from '../resources/messages';

export function enable(req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  // TODO
  res.send('Not implemented yet!');
}

export function disable(req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  // TODO
  res.send('Not implemented yet!');
}

/**
 * Handles ENABLE SCHEDULE request for single server
 */
export function enableServer(req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  const { serverId } = validateInputParameters(req as ApiRequest, res);

  const serverState: ServerState = appState.servers[serverId];
  serverState.isScheduleEnabled = true;
  serverState.onJob.start();
  serverState.offJob.start();

  logger.info(`(enable-server:${serverId}) schedule enabled`);

  res.send(messages.status.okay);
}

/**
  * Handles DISABLE SCHEDULE request for single server
  */
export function disableServer(req: Express.Request, res: TypedResponse<string>, appState: AppState) {
  const { serverId } = validateInputParameters(req as ApiRequest, res);

  const serverState: ServerState = appState.servers[serverId];
  serverState.isScheduleEnabled = false;
  serverState.onJob.stop();
  serverState.offJob.stop();

  logger.info(`(disable-server:${serverId}) schedule disabled`);

  res.send(messages.status.okay);
}
