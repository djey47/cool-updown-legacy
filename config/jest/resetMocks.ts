import globalMocks from '../../config/jest/globalMocks';

const {
    axiosMock, expressResponseMock, nodeFSMock, systemGatewayMock, nodesshMock, pageMock
} = globalMocks;
  
export default function() {
    axiosMock.get.mockReset();
    expressResponseMock.sendMock.mockReset();
    expressResponseMock.statusMock.mockReset();
    nodeFSMock.readFile.mockReset();
    systemGatewayMock.pingMock.mockReset();
    nodesshMock.connect.mockReset();
    nodesshMock.dispose.mockReset();
    nodesshMock.execCommand.mockReset();
    pageMock.generatePage.mockReset();
}
