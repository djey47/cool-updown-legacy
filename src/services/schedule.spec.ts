import globalMocks from '../../config/jest/globalMocks';
import resetMocks from '../../config/jest/resetMocks';
import { disable, enable } from './schedule';

const { expressResponseMock, jobsMock } = globalMocks;

const appState = {
  isScheduleEnabled: true,
  startedAt: new Date('June 12, 2018 13:14:00Z'),
};

const res = {
  status: c => expressResponseMock.statusMock(c),
  send: msg => expressResponseMock.sendMock(msg),
};

const mockJobs = {
  onJob: {
    start: () => jobsMock.onJobStart(),
    stop: () => jobsMock.onJobStop(),
  },
  offJob: {
    start: () => jobsMock.offJobStart(),
    stop: () => jobsMock.offJobStop(),
  },
};

describe('schedule services', () => {
  describe('enable service', () => {
    beforeEach(() => {
      resetMocks();
    });

    it('should set jobs state correctly and generate correct response', () => {
      // given
      const state = {
        ...appState,
        ...mockJobs,
        isScheduleEnabled: false,
      };

      // when
      enable(undefined, res, state);

      // then
      expect(jobsMock.onJobStart).toHaveBeenCalled();
      expect(jobsMock.offJobStart).toHaveBeenCalled();
      expect(state.isScheduleEnabled).toEqual(true);
      expect(expressResponseMock.sendMock).toHaveBeenCalled();
      expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('disable service', () => {
    beforeEach(() => {
      resetMocks();
    });

    it('should set jobs state correctly and generate correct response', () => {
      // given
      const state = {
        ...appState,
        ...mockJobs,
        isScheduleEnabled: true,
      };

      // when
      disable(undefined, res, state);

      // then
      expect(jobsMock.onJobStop).toHaveBeenCalled();
      expect(jobsMock.offJobStop).toHaveBeenCalled();
      expect(state.isScheduleEnabled).toEqual(false);
      expect(expressResponseMock.sendMock).toHaveBeenCalled();
      expect(expressResponseMock.sendMock.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
