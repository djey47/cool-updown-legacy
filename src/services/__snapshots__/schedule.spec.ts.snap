// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schedule services disable service for single server should set jobs state correctly and generate correct response 1`] = `"OK"`;

exports[`schedule services enable service for single server should set jobs state correctly and generate correct response 1`] = `"OK"`;
